server/index.js

```diff
@@ -1,11 +1,522 @@
+});
+
+app.get('/api/cards', (req, res) => {
+  const sql = `select *
+                from "cards";
+              `;
+  db.query(sql)
+    .then(result => {
+      res.status(200).json(result.rows);
+    })
+    .catch(err => {
+      console.error(err);
+      res.status(500).json({
+        error: 'An unexpected error occurred.'
+      });
+    });
+});
+
+app.get('/api/cards/:batchId', (req, res, next) => {
+  const batchId = parseInt(req.params.batchId, 10);
+  if (!Number.isInteger(batchId) || batchId <= 0) {
+    res.status(400).json({
+      error: '"batchId" must be a positive integer'
+    });
+    return;
+  }
+  const sql = `
+    select *
+      from "cards"
+     where "batchId" = $1;
+  `;
+
+  const params = [batchId];
+  db.query(sql, params)
+    .then(result => {
+      const batch = result.rows;
+      if (!batch) {
+        res.status(404).json({
+          error: `Cannot find grade with "batchId" ${batchId}`
+        });
+      } else {
+        res.json(batch);
+      }
+    })
+    .catch(err => {
+      console.error(err);
+      res.status(500).json({
+        error: 'An unexpected error occurred.'
+      });
+    });
+});
+
+app.use(errorMiddleware);
+
+app.patch('/api/cards/:cardId', (req, res) => {
+  const cardId = parseInt(req.params.cardId, 10);
+  if (!Number.isInteger(cardId) || cardId <= 0) {
+    res.status(400).json({
+      error: '"cardId" must be a positive integer'
+    });
+    return;
+  }
+  const { question, answer } = req.body;
+  if (typeof question === 'undefined' || typeof answer === 'undefined') {
+    res.status(400).json({
+      error: 'Question and answer are required fields.'
+    });
+    return;
+  }
+
+  const sql = `update "cards"
+             set "question" = $2,
+                 "answer" = $3
+             where "cardId" = $1
+             returning *;
+             `;
+
+  const params = [cardId, question, answer];
+  db.query(sql, params)
+    .then(result => {
+      const card = result.rows[0];
+      if (!card) {
+        res.status(404).json({
+          error: `Cannot find card with "cardId" ${cardId}`
+        });
+      } else {
+        res.status(200).json(card);
+      }
+    })
+    .catch(err => {
+      console.error(err);
+      res.status(500).json({
+        error: 'An unexpected error occurred.'
+      });
+    });
+});
+
+app.delete('/api/cards/:cardId', (req, res) => {
+  const cardId = parseInt(req.params.cardId, 10);
+  if (!Number.isInteger(cardId) || cardId <= 0) {
+    res.status(400).json({
+      error: '"cardId" must be a positive integer'
+    });
+    return;
+  }
+  const sql = `
+    delete from "cards"
+     where "cardId" = $1
+     returning *;
+  `;
+  const params = [cardId];
+  db.query(sql, params)
+    .then(result => {
+      const card = result.rows[0];
+      if (!card) {
+        res.status(404).json({
+          error: `Cannot find card with "cardId" ${cardId}`
+        });
+      } else {
+        res.status(204).json(card);
```

Comment:
Status code `204` means `No Content` so express won't send anything back. Even though you have `.json(card)` here express will ignore that and send nothing back because of the `204` status code.